### PR TITLE
dev/core#1230 [Ref] Rationalise dedupe code loop.

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Minor code cleanup 

Before
----------------------------------------
Well tested function behaves as expected

After
----------------------------------------
Well tested function behaves as expected, but it also makes more sense

Technical Details
----------------------------------------
We have a function 'dedupePair' which is intended to act on a specific pair
and a loop function which iterates them.

The 'pair actions' and the 'looping actions' are currently jumbled together.

This moves the 'pair actions' to the dedupePair function
and keeps the 'looping actions' in the parent looping function.

Athough I left it out of scope for this PR the api that calls this
function should only call the 'dedupePair' function not the multiple
pair wrapper

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@pfigel now you've done the important ones...  this is just code tidy up but would be good to merge

**Comment by eileenmcnaughton** [comment]
@demeritcowboy if you  are  still  reviewing in this area maybe you could look at  this one - no worries if not it's just a minor tidy up

File: CRM/Dedupe/Merger.php [hunk 1]
```diff
@@ -873,16 +873,17 @@ public static function merge($dupePairs = [], $cacheParams = [], $mode = 'safe',
           unset($dupePairs[$index]);
           continue;
         }
-        CRM_Utils_Hook::merge('flip', $dupes, $dupes['dstID'], $dupes['srcID']);
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: These lines relate to a single pair  - so I moved them into the single pair function

File: CRM/Dedupe/Merger.php [hunk 2]
```diff
@@ -873,16 +873,17 @@ public static function merge($dupePairs = [], $cacheParams = [], $mode = 'safe',
...
-        $mainId = $dupes['dstID'];
-        $otherId = $dupes['srcID'];
-
-        if (!$mainId || !$otherId) {
-          // return error
-          return FALSE;
+        if (($result = self::dedupePair($dupes, $mode, $checkPermissions, $cacheKeyString)) == FALSE) {
+          unset($dupePairs[$index]);
```
Comment (eileenmcnaughton) [inline CRM/Dedupe/Merger.php]: These relate to managing the loop - so I moved  them OUT of the single pair function

**Comment by demeritcowboy** [comment]
Sure - looking...

**Comment by demeritcowboy** [comment]
It's not really my main area and I've never looked at this block of code so it took me a bit to get into it but I was able to do some review.

* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Did a couple simple batch merge tests (which seems to be where this code is used). Results before seemed same as after.
* Developer standards
    * [`r-tech`] __PASS__
    * [`r-code`] __Undecided__
      * In the original, it would exit the task completely at line 882. Now it just unsets something at line 877 and continues. Is the change intended? I couldn't think where this would come up.
      * Also, line 876 in the new file is a loose == instead of strict ===.
      * Since I'm out of my area on this I'll just mention it and defer - I don't see dedupePair() used in core outside of this class, but if it's used or subclassed somewhere I'm not seeing it, then the signature change is a change (as opposed to a deprecation).
    * [`r-maint`] __N/A__
    * [`r-test`] __PASS__

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I fixed the non strict comparison - strict is better practice.

WRT the other - I really can't think of when it would be hit - it seems to be a check for something that should never be true. However, the 2 most valid behaviours seem to me to be 'skip this pair & continue'  or throw an exception. 

In the 2 places this function is called from the batch merge expects a returned array & will get kinda weirded out if it doesn't get one. The api does cope with the 'FALSE' but would also cope with an exception. I think the handling might validation in the wrong place for the api call  (any opinions @pfigel @JKingsnorth ?)

**Comment by mattwire** [comment]
@pfigel @bjendres @JKingsnorth Please can one of you comment on the final comment from @eileenmcnaughton and confirm if the issues identified by @demeritcowboy could be an issue?

**Comment by seamuslee001** [comment]
I think the change to unsetting makes sense in my opinion.  i think it feels more correct to keep trying to go around the loop if we hit any issues

**Comment by seamuslee001** [comment]
Added the merge ready flag to give others any chance to comment but plan to merge in a couple of days time

**Comment by eileenmcnaughton** [comment]
Merging per @seamuslee001 'in a couple of days time' comment

**Pull request merged**